backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/FileSetManager.java

```diff
@@ -60,16 +132,72 @@ CompletableFuture<NamedFileSet> restore(
     return CompletableFutureUtils.mapAsync(
             fileSet.files().entrySet(),
             Entry::getKey,
-            namedFile -> restoreFile(sourcePrefix, targetFolder, namedFile.getKey()))
+            namedFile ->
+                restoreFile(sourcePrefix, targetFolder, namedFile.getKey(), namedFile.getValue()))
         .thenApply(NamedFileSetImpl::new);
   }
 
   private CompletableFuture<Path> restoreFile(
-      final String sourcePrefix, final Path targetFolder, final String fileName) {
+      final String sourcePrefix,
+      final Path targetFolder,
+      final String fileName,
+      final FileMetadata metadata) {
+    final var compressionAlgorithm = metadata.compressionAlgorithm();
+    if (compressionAlgorithm.isPresent()) {
+      final var decompressed = targetFolder.resolve(fileName);
+      LOG.trace(
+          "Restoring compressed file {} from prefix {} to {}",
+          fileName,
+          sourcePrefix,
+          targetFolder);
+      try {
+        final var compressed = Files.createTempFile(TMP_DECOMPRESSION_PREFIX, null);
+        return client
+            .getObject(
+                req -> req.bucket(config.bucketName()).key(sourcePrefix + fileName),
+                AsyncResponseTransformer.toFile(
+                    compressed,
+                    cfg ->
+                        cfg.fileWriteOption(FileWriteOption.CREATE_OR_REPLACE_EXISTING)
+                            .failureBehavior(FailureBehavior.DELETE)))
+            .thenApplyAsync(
+                response -> decompressFile(compressed, decompressed, compressionAlgorithm.get()));
+
+      } catch (final IOException e) {
+        throw new UncheckedIOException(e);
+      }
+    }
+
     LOG.trace("Restoring file {} from prefix {} to {}", fileName, sourcePrefix, targetFolder);
     final var path = targetFolder.resolve(fileName);
     return client
         .getObject(req -> req.bucket(config.bucketName()).key(sourcePrefix + fileName), path)
         .thenApply(response -> path);
   }
+
+  private Path decompressFile(
+      final Path compressed, final Path decompressed, final String algorithm) {
+    try (final var input = new BufferedInputStream(Files.newInputStream(compressed));
+        final var output = new BufferedOutputStream(Files.newOutputStream(decompressed));
+        final var decompressedOutput =
+            new CompressorStreamFactory().createCompressorInputStream(algorithm, input)) {
+      IOUtils.copy(decompressedOutput, output);
+      if (LOG.isTraceEnabled()) {
```

Comment:
:wrench: Why would you add this check, but not in other places? I don't think this is required.

Reply:
:+1: